minithesis.py

```diff
@@ -328,21 +387,26 @@ def produce(test_case):
             result.append(test_case.any(elements))
         return result
 
-    return Possibility(produce, name=f"lists({elements.name})")
+    return Possibility[List[T]](produce, name=f"lists({elements.name})")
 
 
-def just(value):
+def just(value: T) -> Possibility[T]:
     """Only ``value`` is possible."""
-    return Possibility(lambda tc: value, name=f"just({value})")
+    return Possibility[T](lambda tc: value, name=f"just({value})")
 
 
-def nothing():
+def nothing() -> Possibility[NoReturn]:
     """No possible values. i.e. Any call to ``any`` will reject
     the test case."""
-    return Possibility(lambda tc: tc.reject())
 
+    def produce(tc: TestCase) -> NoReturn:
+        tc.reject()
+
+    return Possibility(produce)
 
-def mix_of(*possibilities):
+
+# XXX This signature requires PEP 646
+def mix_of(*possibilities: Possibility[Any]) -> Possibility[Any]:
```

Comment:
Well, `Any` just shuts up the type checker. If I change the signature to `Possibility[T]` there's a slight problem because `nothing()` isn't a valid `Possibility[T]` -- but this is due to the hacky nature of `NoReturn`; if `NoReturn` was a proper bottom type it would be a subtype of all other types, so I'll just add a cast.

Reply:
Hmm. I don't know how well the type checker handles this, but could you make it so `nothing() -> Possibility[T]`? `nothing()` is logically a `Possibility[T]` for any value of `T`, which would solve that particular problem if the type parameter is inferred appropriately when it's used.

Reply:
But that's not how type variables work... The `T` used in the signature of `nothing()` would be unrelated to the `T` used in `mix_of()`, and in fact we'd get an error if we wrote
```
def nothing() -> Possibility[T]:  # E: TypeVar "T" appears only once in generic function signaturePylancereportInvalidTypeVarUse
```
so I think the cast is the best we can do, until the type system changes and makes `NoReturn` a proper "bottom" type.

Reply:
Right, whether we'd get that error was basically what I was asking. In some languages this will type check fine, because you can infer the type parameter from the point of use, but I guess the Python type checkers don't do that much inference. In that case yeah I agree the cast is probably the best one can do.